kimchi/src/circuits/lookup/lookups.rs

```diff
@@ -550,3 +550,50 @@ fn lookup_pattern_constants_correct() {
         assert_eq!((pat, pat.max_joint_size()), (pat, max_joint_size as u32));
     }
 }
+
+#[cfg(feature = "wasm_types")]
+pub mod wasm {
+    use super::*;
+
+    #[wasm_bindgen::prelude::wasm_bindgen]
+    impl LookupPatterns {
+        #[wasm_bindgen::prelude::wasm_bindgen]
+        pub fn new(xor: bool, lookup: bool, range_check: bool, foreign_field_mul: bool) -> Self {
+            Self {
+                xor,
+                lookup,
+                range_check,
+                foreign_field_mul,
+            }
+        }
+    }
+
+    #[wasm_bindgen::prelude::wasm_bindgen]
+    impl LookupFeatures {
+        #[wasm_bindgen::prelude::wasm_bindgen]
+        pub fn new(
+            patterns: LookupPatterns,
+            joint_lookup_used: bool,
+            uses_runtime_tables: bool,
+        ) -> Self {
+            Self {
+                patterns,
+                joint_lookup_used,
+                uses_runtime_tables,
+            }
+        }
+    }
+
+    #[wasm_bindgen::prelude::wasm_bindgen]
+    impl LookupInfo {
+        // Not quite sure about the actual types for max_per_row and max_joint_size
```

Comment:
they're JS `number`s. your code should be fine in that it preserves any reasonably large value

Reply:
Hum so they're `f64` then?

Reply:
Hmm, I think internally in JS, `number` is either `f64` or `i32`, depending on the circumstance, but when passed to Wasm will be converted to the type matching the input type of the generated Wasm function. In this case definitely to a 32-bit integer, `i32`. (Wasm types don't distinguish between `u32` and `i32`, instead the individual instructions treat them as one or the other)

Reply:
JS numbers are clearly double

Reply:
that article also mentions that these doubles are converted to int32 for some operations:

> JavaScript has some lower-level functions that deal with the binary encoding of integer numbers, most notably [bitwise operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators#bitwise_shift_operators) and [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) objects. Bitwise operators always convert the operands to 32-bit integers.

what the article doesn't mention is that there are known ways of writing code that the compiler will take as hint that it should use the int32 representation for several operations in a row without converting back to double